onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -2139,11 +2141,83 @@ Example 4:
         }
 
         if (ctx.getNumOutputs() > 2) {
-          auto saved_inv_std_var_shape = ctx.getOutputType(2)->mutable_tensor_type()->mutable_shape();
-          saved_inv_std_var_shape->CopyFrom(input_shape);
-          saved_inv_std_var_shape->mutable_dim(static_cast<int>(axis))->set_dim_value(1);
+          auto saved_inv_std_dev_shape = ctx.getOutputType(2)->mutable_tensor_type()->mutable_shape();
+          saved_inv_std_dev_shape->CopyFrom(input_shape);
+          saved_inv_std_dev_shape->mutable_dim(static_cast<int>(axis))->set_dim_value(1);
         }
-      });
+      })
+      .SetContextDependentFunctionBodyBuilder(
+          [](const FunctionBodyBuildContext& ctx, const OpSchema& schema, FunctionProto& functionProto) {
+            // LayerNormalization <axis, epsilon, stash_type> (X, Scale, B) => (Y, Mean?, InvStdDev?)
+
+            auto* tp = ctx.getInputType(0);
+            if ((tp == nullptr) || (!tp->has_tensor_type()))
+              return false;
+            int64_t T = tp->tensor_type().elem_type();
+
+            auto type_attr = ctx.getAttribute("stash_type");
+            int64_t U = (type_attr != nullptr) ? type_attr->i() : static_cast<int64_t>(ONNX_NAMESPACE::TensorProto_DataType_FLOAT);
+            if ((U != ONNX_NAMESPACE::TensorProto_DataType_FLOAT) && (U != ONNX_NAMESPACE::TensorProto_DataType_BFLOAT16))
```

Comment:
>((U != ONNX_NAMESPACE::TensorProto_DataType_FLOAT) && (U != ONNX_NAMESPACE::TensorProto_DataType_BFLOAT16 [](http://example.com/codeflow?start=15&length=105)

Could you add some comments about why float16 is not supported?

Reply:
Good question. I am just following the existing schema, which allows only float and bfloat16, not changing it. Do you think we should change the schema to allow float16 also? But note that this is the type of mean/std-dev, which is allowed to be more precise than the inputs, which may be low-precision. May be we are not using low-precision for mean/std-dev currently?